jni/src/faiss_index_service.cpp

```diff
@@ -57,6 +57,103 @@ void SetExtraParameters(knn_jni::JNIUtilInterface * jniUtil, JNIEnv *env,
 
 IndexService::IndexService(std::unique_ptr<FaissMethods> faissMethods) : faissMethods(std::move(faissMethods)) {}
 
+jlong IndexService::initIndex(
+        knn_jni::JNIUtilInterface * jniUtil,
+        JNIEnv * env,
+        faiss::MetricType metric,
+        std::string indexDescription,
+        int dim,
+        int numVectors,
+        int threadCount,
+        std::unordered_map<std::string, jobject> parameters
+    ) {
+    // Create index using Faiss factory method
+    std::unique_ptr<faiss::Index> indexWriter(faissMethods->indexFactory(dim, indexDescription.c_str(), metric));
+
+    // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
+    if(threadCount != 0) {
+        omp_set_num_threads(threadCount);
+    }
+
+    // Add extra parameters that cant be configured with the index factory
+    SetExtraParameters<faiss::Index, faiss::IndexIVF, faiss::IndexHNSW>(jniUtil, env, parameters, indexWriter.get());
+
+    // Check that the index does not need to be trained
+    if(!indexWriter->is_trained) {
+        throw std::runtime_error("Index is not trained");
+    }
+
+    // Add vectors
+    std::unique_ptr<faiss::IndexIDMap> idMap (faissMethods->indexIdMap(indexWriter.release()));
```

Comment:
Nit. If we are going to keep the memory, I think smart pointer only makes it complex. We can simply use plain pointer. 

Here, it would be better to set `idMap->own_fields = true;` ihere nstead of we call `delete idMap-index` later.

Reply:
I talked with @navneet1v about this, but own_fields doesn't work as intended. Don't know if this is a bug in faiss or a bug in how we initialize an index.

Reply:
The smart pointer is a good idea, though, since we don't have to worry about releasing memory for all possible exceptions.

Reply:
Yes , Smart Pointer releases memory automatically when it goes out of scope, we don't have to write delete idmap, it's automatically taken care.
https://fastercapital.com/content/Smart-Pointers-in-C----Managing-Memory-Safely.html#:~:text=Instead%20of%20returning%20a%20raw,overall%20safety%20of%20your%20code.

Reply:
Resolved since this is the best fix for freeing index storage